server/metadata/metadata.go

```diff
@@ -53,29 +78,77 @@ func (m *metadataSubspace) getMetadata(ctx context.Context, tx transaction.Tx, i
 
 	var row kv.KeyValue
 	if !it.Next(&row) {
-		return nil, it.Err()
+		if it.Err() != nil {
+			return nil, it.Err()
+		}
+
+		return nil, errors.ErrNotFound
 	}
 
-	log.Debug().Str("type", string(m.SubspaceName)).Str("key", key.String()).Str("value", string(row.Data.RawData)).Msgf("reading metadata succeed")
+	log.Debug().Str("type", string(m.SubspaceName)).Str("key", key.String()).
+		Str("value", string(row.Data.RawData)).Msgf("reading metadata succeed")
 
-	return row.Data.RawData, nil
+	return row.Data, nil
 }
 
-func (m *metadataSubspace) updateMetadata(ctx context.Context, tx transaction.Tx, invalidArgs error, key keys.Key, payload []byte) error {
+func (m *metadataSubspace) getMetadata(ctx context.Context, tx transaction.Tx, invalidArgs error, key keys.Key,
+	metadata any,
+) error {
 	if invalidArgs != nil {
 		return invalidArgs
 	}
 
-	_, err := tx.Update(ctx, key, func(data *internal.TableData) (*internal.TableData, error) {
-		return internal.NewTableData(payload), nil
+	payload, err := m.getPayload(ctx, tx, nil, key)
+	if err != nil {
+		return err
+	}
+
+	if payload == nil {
+		return errors.ErrNotFound
+	}
+
+	if err = jsoniter.Unmarshal(payload.RawData, metadata); ulog.E(err) {
+		return errors.Internal("failed to unmarshal metadata")
+	}
+
+	return nil
+}
+
+func (m *metadataSubspace) updatePayload(ctx context.Context, tx transaction.Tx, invalidArgs error, key keys.Key,
+	ver int32, payload []byte,
+) error {
+	if invalidArgs != nil {
+		return invalidArgs
+	}
+
+	_, err := tx.Update(ctx, key, func(_ *internal.TableData) (*internal.TableData, error) {
```

Comment:
Would it not be better to make this a replace? Do we need to read the key before updating it?